README.md
A flutter widget that provides a circular widget that allows you make customizations in its properties and build pretty animations using it.

## Features

Creates a circular visual widget that you can specify the size, stroke width, colors and desired values to fill determined angle of circle. Supports animations like Tweens to change the current value of fill and much more.

Start             |  Middle             |  End
:-------------------------:|:-------------------------:|:-------------------------:
![Start](./images/start.png)  |  ![Middle](./images/middle.png)  |  ![End](./images/end.png)

## How to install
Add this dependency to your pubspec.yml:

```yaml
dependencies:
  circle_gadget: ^1.0.0
```

Now in your Dart code, you can use:

```dart
import 'package:circle_gadget/circle_gadget.dart';
```

## Usage

The `circle_gadget` package itself is very simple to use just create a widget as stateless or statefull

```dart
CircularGadgetWidget(
    width: 200,
    height: 200,
    strokeColor: Theme.of(context).primaryColor.withAlpha(40),
    strokeValueColor: Theme.of(context).primaryColor,
    centerColor: Colors.grey.withAlpha(10),
    strokeWidth: 8,
    min: 0.0,
    max: 180.0,
    value: 100.0,
)
```

See the full example [circular animation](https://github.com/MarioJunio/flutter-circle-gadget/tree/main/lib/examples).

## Widget properties
**width**
*double*\
Change the circle size width

**height**
*double*\
Change the circle size height

**strokeWidth**
*double*\
Change the circle stroke size width

**strokeColor**
*Color*\
Change the circle stroke color

```dart
strokeColor: Colors.blue.withAlpha(50)
```

**strokeValueColor**
*Color*\
Change the circle stroke color value

```dart
strokeColor: Colors.blue
```

**min**
*double*\
Set the min value that circle waits

**max**
*double*\
Set the max value that circle waits

**centerColor**
*Color*\
Change the center color of circle

```dart
strokeColor: Colors.white.withAlpha(30)
```


CHANGELOG.md
## 1.0.0

* Circular Gadget

## 1.0.1

* Updates in README.md

lib/circle_gadget.dart
library circle_gadget;

import 'package:circle_gadget/renders/circular_gadget_render_object.dart';
import 'package:flutter/material.dart';

class CircularGadgetWidget extends StatelessWidget {
  const CircularGadgetWidget({
    super.key,
    required this.width,
    required this.height,
    required this.min,
    required this.max,
    required this.value,
    required this.strokeWidth,
    required this.strokeColor,
    required this.strokeValueColor,
    required this.centerColor,
  });

  final double width, height;
  final double min, max, value;
  final double strokeWidth;
  final Color strokeColor;
  final Color strokeValueColor;
  final Color centerColor;

  @override
  Widget build(BuildContext context) {
    return SizedBox.fromSize(
      size: Size(width, height),
      child: CircularGadgetRenderObject(
        centerColor: centerColor,
        strokeColor: strokeColor,
        strokeValueColor: strokeValueColor,
        strokeWidth: strokeWidth,
        min: min,
        max: max,
        value: value,
        key: key,
      ),
    );
  }
}


lib/renders/circular_gadget_render_object.dart
import 'package:circle_gadget/renders/circular_gadget_render.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

class CircularGadgetRenderObject extends LeafRenderObjectWidget {
  const CircularGadgetRenderObject({
    super.key,
    required this.strokeWidth,
    required this.strokeColor,
    required this.strokeValueColor,
    required this.centerColor,
    required this.min,
    required this.max,
    required this.value,
  });

  final double strokeWidth;
  final Color strokeColor;
  final Color strokeValueColor;
  final Color centerColor;
  final double min, max, value;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return CircularGadgetRender(
      centerColor: centerColor,
      strokeColor: strokeColor,
      strokeValueColor: strokeValueColor,
      strokeWidth: strokeWidth,
      max: max,
      min: min,
      value: value,
    );
  }

  @override
  void updateRenderObject(
      BuildContext context, covariant CircularGadgetRender renderObject) {
    renderObject.strokeColor = strokeColor;
    renderObject.strokeValueColor = strokeValueColor;
    renderObject.centerColor = centerColor;
    renderObject.min = min;
    renderObject.max = max;
    renderObject.value = value;
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);

    properties.add(ColorProperty("strokeColor", strokeColor));
    properties.add(ColorProperty("strokeValueColor", strokeValueColor));
    properties.add(ColorProperty("centerColor", centerColor));
    properties.add(DoubleProperty("strokeWidth", strokeWidth));
    properties.add(DoubleProperty("min", min));
    properties.add(DoubleProperty("max", max));
    properties.add(DoubleProperty("value", value));
  }
}


lib/examples/example_page.dart
import 'package:circle_gadget/circle_gadget.dart';
import 'package:flutter/material.dart';

class ExamplePage extends StatefulWidget {
  const ExamplePage({Key? key}) : super(key: key);

  @override
  State<ExamplePage> createState() => _ExamplePageState();
}

class _ExamplePageState extends State<ExamplePage>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _gadgetValueTween;

  @override
  void initState() {
    super.initState();

    _animationController = AnimationController(
      vsync: this,
      duration: const Duration(seconds: 2),
    );

    _gadgetValueTween = Tween<double>(
      begin: 0,
      end: 180,
    ).animate(
      CurvedAnimation(
        parent: _animationController,
        curve: Curves.easeInOut,
      ),
    );

    // ignore: avoid_single_cascade_in_expression_statements
    _animationController
      ..addStatusListener((status) {
        setState(() {});
      });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          _startAnimationBtn(context),
          _circleGadget,
        ],
      ),
    );
  }

  Widget get _circleGadget => Padding(
        padding: const EdgeInsets.symmetric(vertical: 32),
        child: Center(
          child: _animatedBuilder(context),
        ),
      );

  AnimatedBuilder _animatedBuilder(context) => AnimatedBuilder(
        animation: _animationController,
        builder: (context, child) {
          const double width = 200;
          const double height = 200;

          return SizedBox.fromSize(
            size: const Size(width, height),
            child: Stack(
              children: [
                Positioned.fill(
                  child: Align(
                    alignment: Alignment.center,
                    child: Text(
                      "${_gadgetValueTween.value.toInt()} km/h",
                      style: Theme.of(context).textTheme.bodyLarge!.copyWith(
                            color: Theme.of(context).primaryColor,
                            fontWeight: FontWeight.w700,
                            fontSize: 18,
                          ),
                    ),
                  ),
                ),
                CircularGadgetWidget(
                  strokeColor: Theme.of(context).primaryColor.withAlpha(40),
                  strokeValueColor: Theme.of(context).primaryColor,
                  centerColor: Colors.grey.withAlpha(10),
                  strokeWidth: 8,
                  min: 0.0,
                  max: 180.0,
                  value: _gadgetValueTween.value,
                  width: width,
                  height: height,
                ),
              ],
            ),
          );
        },
      );

  Widget _startAnimationBtn(context) => Padding(
        padding: const EdgeInsets.symmetric(horizontal: 32),
        child: ElevatedButton(
          onPressed: () {
            if (_animationController.isCompleted) {
              _animationController.reverse();
            } else {
              _animationController.forward();
            }
          },
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              _animationController.isCompleted ? "Reverse" : "Start",
              style: Theme.of(context).textTheme.labelMedium,
            ),
          ),
        ),
      );
}
